### ISSUE TITLE
2D directional Light does not behave as expected in Godot 4

### ISSUE THREAD

**Issue by DocoDan** [godotengine/godot#48757]
**Godot version:**
4.0.dev.custom_build.2295d3eba

**OS/device including version:**
Ubuntu 20.04.2 LTS

**Issue description:**
It seems only possible to either affect all layers by a DirectionalLight2D or none of them. 
This behavior is controlled by setting or not setting the Layer Min value
of the DirectionalLigh2Ds Range-Property to either 0 or any other value. 
If I exclude 0 of the Min and Max range, none of the layers is affected, otherwise every layer is affected. 
E.g. I do not want my Backgroundtexture to be affected, but no matter which Light-Mask-Layer-Combination 
I try for this node, it either is affected by the DirectionalLight2D or it is not. 
And this seems to solely depend of the RangeProperty-Settings of the DirectionalLight2D node.
I would expect the Texture to be affected only by the light node if the light masks are matching, but this is not the case.

**Steps to reproduce:**
Create a new project and add a Sprite and a DirectionalLight2D node. Under Visibility choose different options for the Light Mask Layer of your Sprite Node. The DirectionalLight2D will always affect your Sprite-Node, no matter if its Light Mask matches the Item Cull mask of the DirectionalLight2D or not.

**Minimal reproduction project:**
[2dDirectionalLightGodot4.zip](https://github.com/godotengine/godot/files/6488948/2dDirectionalLightGodot4.zip)

**Comment by nabfrew** [comment]
I'm having this issue too. It does not respect the item_cull_mask.

**Comment by clayjohn** [comment]
Thinking about this a little more, I think I have an idea for how it could be implemented. Right now the limiting factor is we only have 128 bytes available per object. So there isn't really room to store a 32 bit mask. 

However, I have plans to change the lighting implementation in OpenGL to make the light list uniform across the entire batch (meaning that having lights will break batches). This would allow us to take the light information out of the per-object data and move it into the per-batch data and free up some room for more information. 

For Vulkan, we would have to implement batching and then do the same optimization. 

Before committing to this solution I would need to test it out and see if it actually leads to an increase in performance for 2D lights in the OpenGL backend.

**Comment by aoetiker** [comment]
> Thinking about this a little more, I think I have an idea for how it could be implemented. Right now the limiting factor is we only have 128 bytes available per object. So there isn't really room to store a 32 bit mask.
> 
> However, I have plans to change the lighting implementation in OpenGL to make the light list uniform across the entire batch (meaning that having lights will break batches). This would allow us to take the light information out of the per-object data and move it into the per-batch data and free up some room for more information.
> 
> For Vulkan, we would have to implement batching and then do the same optimization.
> 
> Before committing to this solution I would need to test it out and see if it actually leads to an increase in performance for 2D lights in the OpenGL backend.

Has anything happened here in the meantime? the feature would be fantastic. I don't want my parallax background lit

**Comment by valkyrienyanko** [comment]
I found a temporary work-a-round for this, although this may not work for you.

Create a new PointLight2D node and rename it to DirectionalLight2D. Set the texture in the PointLight2D to a new GradientTexture2D and make the gradient completely white. Set the texture width to something like 1024 by 1024 pixels. Set the scale of the PointLight2D node to (100, 100). Set blend mode to subtract and energy to 1.

Attach this node to your player. Now the light will follow your player wherever it goes and will act as a "DirectionalLight2D".

The cull masks will work as expected. As you can see below this purple blueish circle is on a different light mask so it is not effected by the point light attached to the player. Although the player is on a light mask that the quote unquote "DirectionalLight2D" is using, so of course you won't be able to see the player which is inside the circle.

![Untitled](https://github.com/godotengine/godot/assets/6277739/6d598689-db72-4da9-887b-02128f7f5302)

I hope this helps anyone out.

**Comment by aoetiker** [comment]
> **TL;DR: PointLight2D is pretty much a working version of DirectionalLight2D. (Cull masks work 100% in PointLight2D for some reason)**
> 
> This workaround can be used in the meantime until this issue is fixed.
> 
> Create a new PointLight2D node and rename it to DirectionalLight2D. Set the texture in the PointLight2D to a new GradientTexture2D and make the gradient completely white. Set the texture width to something like 1024 by 1024 pixels and set the scale of the PointLight2D node to (100, 100) so the point light for sure covers the entire game screen. Set blend mode to subtract and energy to 1.
> 
> Attach this node to your player. Now the light will follow your player wherever it goes and will act as a "DirectionalLight2D".
> 
> _Edit: The light node should be attached to a camera that is set to follow the player via script. This way when the player dies, the directional light does not die._
> 
> The cull masks will work as expected.
> 
> As a bonus you can set the "DirectionalLight2D" to a scene and copy it over to your main level scene. Lock it so you can't select it. And set it to "Editor Only" so you can preview your lighting effects. Also you can disable the player point light in the inspector and enable it when the game starts through script so you can see your player when editing it in the editor.
> 
> I hope this helps someone out.

Unfortunately, that makes everything just black. In my case, I want the background texture not be affected by lighting.

I found the following solution:

1. Add a CanvasItemMaterial to the Background Image
2. Set Blend Mode to Mix
3. Set Light Mode to Unshaded

With that, the background is not affected by Lighting and I can change the tint by using Modulate if needed.

But it would be great, if that bug will be fixed soon...

**Comment by DarinDev1000** [comment]
I am having the same issue. It took me a bit to figure out that the Item Cull Mask was doing nothing.

**Comment by DarinDev1000** [comment]
> > **TL;DR: PointLight2D is pretty much a working version of DirectionalLight2D. (Cull masks work 100% in PointLight2D for some reason)**
> > This workaround can be used in the meantime until this issue is fixed.
> > Create a new PointLight2D node and rename it to DirectionalLight2D. Set the texture in the PointLight2D to a new GradientTexture2D and make the gradient completely white. Set the texture width to something like 1024 by 1024 pixels and set the scale of the PointLight2D node to (100, 100) so the point light for sure covers the entire game screen. Set blend mode to subtract and energy to 1.
> > Attach this node to your player. Now the light will follow your player wherever it goes and will act as a "DirectionalLight2D".
> > _Edit: The light node should be attached to a camera that is set to follow the player via script. This way when the player dies, the directional light does not die._
> > The cull masks will work as expected.
> > As a bonus you can set the "DirectionalLight2D" to a scene and copy it over to your main level scene. Lock it so you can't select it. And set it to "Editor Only" so you can preview your lighting effects. Also you can disable the player point light in the inspector and enable it when the game starts through script so you can see your player when editing it in the editor.
> > I hope this helps someone out.
> 
> Unfortunately, that makes everything just black. In my case, I want the background texture not be affected by lighting.
> 
> I found the following solution:
> 
> 1. Add a CanvasItemMaterial to the Background Image
> 2. Set Blend Mode to Mix
> 3. Set Light Mode to Unshaded
> 
> With that, the background is not affected by Lighting and I can change the tint by using Modulate if needed.
> 
> But it would be great, if that bug will be fixed soon...

Thank you! This works for now

**Comment by Scraprogrammer** [comment]
In my game, I don't have a camera and just attach a light node to the side of the screen. In this case, if screen resizes are to be considered - the PointLight2D workaround becomes hard to use, since the light position and gradient texture size must be dynamically re-calculated on resize to try and preserve the same shading effect, no matter the screen size. I was not able to achieve such calculation...  

For now the workaround that works for me is to keep using DirectionalLight2D, but attach a material with `light_mode = 1` to all UI elements that the light should not affect. I do not enjoy this workaround... I wonder if there is some way to set the default `light_mode` to be 1 for all nodes? That way I could reverse my workaround to add material with `light_mode = 0` only to nodes that need shading...

But getting a proper fix to make the mask work would be the best approach still, hope it arrives at some point )

**Comment by iocutus** [comment]
The PointLight2D "solution" is not a solution. What if I want global illumination with cull masks that doesn't follow the player? The PointLight2D, no matter how big, does not cast shadows the way a DirectionalLight2D would. There needs to be a solution that allows for global illumination that respects cull masks, so layers do not cast shadows over layers that are on top of them.

**Comment by aoetiker** [comment]
> The PointLight2D "solution" is not a solution. What if I want global illumination with cull masks that doesn't follow the player? The PointLight2D, no matter how big, does not cast shadows the way a DirectionalLight2D would. There needs to be a solution that allows for global illumination that respects cull masks, so layers do not cast shadows over layers that are on top of them.

...still hoping that it will be fixed soon...

**Comment by Calinou** [comment]
@aoetiker Please don't bump issues without contributing significant new information. Use the :+1: reaction button on the first post instead.
